### DISCUSSION TITLE
`git test fix`, `git test run --search`

### DISCUSSION THREAD

**Discussion by arxanas** [arxanas/git-branchless#803] [Show and tell]
[`git test`](https://github.com/arxanas/git-branchless/wiki/Command:-git-test) has several exciting new features in the upcoming v0.7! Your feedback is appreciated, especially as we introduce a similar `jj run` command to [Jujutsu](https://github.com/martinvonz/jj).

# Fixing

One common task is to apply formatters or linters to an entire stack of commits. This can be tedious to do by hand, especially because formatting changes tend to cause lots of merge conflicts. [`git test fix`](https://github.com/arxanas/git-branchless/wiki/Command:-git-test#fixing-formatting-and-linting-issues) addresses this use-case while guaranteeing no merge conflicts. It also caches results and can be parallelized with the `--jobs` option.

Here's a demo of running rustfmt on all of the commits in a stack:

https://user-images.githubusercontent.com/454057/219904589-79657aed-9356-4f87-a0e4-bdcfbe691621.mov

# Searching

`git test run` implements a search feature [with the `--search <strategy>` option](https://github.com/arxanas/git-branchless/wiki/Command:-git-test#searching). (For the common case of doing a binary search, you can also pass `-b`/`--bisect` as a shorthand.)

Here's a demo (where the test command is implicitly `cargo clippy`):

https://user-images.githubusercontent.com/454057/219905696-400f9086-6cbd-4cc1-a972-b40e6c710e13.mov

And here's another demo where the search is parallelized with the `--jobs` option:

https://user-images.githubusercontent.com/454057/219905728-8bfb1579-e72d-4f6f-9a40-a0cab8b080cc.mov

In comparison to `git bisect`:

- Makes it easier to start the search command on a given range, without manually identifying initial good or bad commits.
  - Actually, it doesn't require that there even exist any good or bad commits; it will just report the first good and bad commits, if any.
- Also supports linear and reverse-linear searching, which are occasionally useful when it's expensive to switch to distant commits and test them.
- Caches the results of the search command, so you can cancel and resume the search later, or expand/contract the search range.
- Can run the search command in parallel on multiple commits.
- Supports searching a DAG (instead of just a linear range of commits) and returning the lower bound of failing commits (instead of just a single failing commit), which can be useful when searching certain graphs with merge commits.
- It's not shown here, but you can still do an interactive bisection by passing the `-i`/`--interactive` flag.

**Comment by LemmingAvalanche** [comment]
Kudos on `git test fix`! Running formatting commands on the snapshots/trees seems so obvious now in hindsight, but I might have never thought of it myself. I think this command will be useful for me for years to come.
